src/charm.py

```diff
@@ -448,6 +506,30 @@ def is_db_service_ready(self) -> bool:
         with MongoDBConnection(self.mongodb_config, "localhost", direct=True) as direct_mongo:
             return direct_mongo.is_ready
 
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Listen to changes in application configuration.
+
+        To prevent a user from migrating a cluster, and causing the component to become
+        unresponsive therefore causing a cluster failure, error the component. This prevents it
+        from executing other hooks with a new role.
+        """
+        if self.is_role_changed():
```

Comment:
This assumes that the role has been set at deploy time. In other words, we cannot go from replication (default) to sharding

Reply:
no we cannot - this isn't something we yet support. happy to chat about making this an epic in the future

Reply:
Role management is coming... spec: DA116